src/V3Randomize.cpp

```diff
@@ -464,7 +583,74 @@ class RandomizeMarkVisitor final : public VNVisitor {
         // of type AstLambdaArgRef. They are randomized too.
         const bool randObject = nodep->fromp()->user1() || VN_IS(nodep->fromp(), LambdaArgRef);
         nodep->user1(randObject && nodep->varp()->rand().isRandomizable());
-        nodep->user2p(m_modp);
+
+        if (m_withp) {
+            AstNode* backp = m_withp;
+            while (backp->backp()) {
+                if (VN_IS(backp, MethodCall)) {
+                    AstClassRefDType* classdtype = VN_CAST(
+                        VN_AS(backp, MethodCall)->fromp()->dtypep()->skipRefp(), ClassRefDType);
+                    nodep->user2p(classdtype->classp());
+                    break;
+                }
+                backp = backp->backp();
+            }
+        } else {
+            nodep->user2p(m_modp);
+        }
+        if (randObject && nodep->varp()
+            && nodep->varp()->rand().isRandomizable()) {  // Process global constraints
+            if (m_classp && m_classp->user1() == IS_RANDOMIZED) {
+                m_classp->user1(IS_RANDOMIZED_GLOBAL);
+            }
+            // Mark the entire nested chain as participating in global constraints
+            // For foo.in.val, this marks: foo, foo.in, and foo.in.val
+            if (VN_IS(nodep->fromp(), VarRef) || VN_IS(nodep->fromp(), MemberSel)) {
+                markNestedGlobalConstrainedRecurse(nodep->fromp());
+            }
+
+            // Global constraint processing algorithm:
+            // 1. Detect globally constrained object variables in randomized classes
+            // 2. Clone constraint trees from the constrained object's class
+            // 3. Rename cloned constraints with object prefix (obj.var format)
+            // 4. Insert cloned constraints into current class for solver processing
+            // 5. Use basic randomization for non-constrained variables to avoid recursion
+
+            // Extract and validate components early to avoid repeated type checks
+            AstVarRef* const varRefp = VN_CAST(nodep->fromp(), VarRef);
+            if (!varRefp || !varRefp->varp()) return;
+
+            const AstClassRefDType* const classRefp
+                = VN_CAST(varRefp->dtypep()->skipRefp(), ClassRefDType);
+            UASSERT(classRefp && classRefp->classp(),
+                    "Global constraint variable must have valid class type");
+
+            if (nodep->user1() && varRefp->varp()->globalConstrained()) {
+                AstClass* gConsClass = classRefp->classp();
+                AstVar* const objVar = varRefp->varp();  // The object variable (e.g., obj1, obj2)
+
+                // Process per-variable (object instance), not per-class
+                // This allows multiple objects of the same class (e.g., obj1 and obj2 of type Sub)
+                if (m_processedVars.find(objVar) == m_processedVars.end()) {
```

Comment:
```suggestion
                if (m_processedVars.insert(objVar).second) {
```

Then delete the `.insert` at the end of the if. This avoid having to lookup the map twice.

(That is unless the `if` body is recursive and re-enters the same code, seems it doesn't)